### DISCUSSION TITLE
preparation of the scada for test/production operation

### DISCUSSION THREAD

**Discussion by joyshmitz** [riclolsen/json-scada#17] [Q&A]
what features are there in launching scads into test operation, followed by production operation:
1. for example clearing the database from demo data ...

**Comment by riclolsen** [comment]
To delete demo data before configuring new tags, connect to mongodb using the mongo shell (mongo.exe) and

    use json_scada
    db.realtimeData.deleteMany({_id: {$gt: 0}})

**Reply by joyshmitz** [reply]
and in postgres (timescaledb) how to properly delete old data?

**Reply by riclolsen** [reply]
Just delete all data.

    psql -U postgres -h 127.0.0.1 json_scada
    delete from hist;
    delete from realtime_data;

**Reply by joyshmitz** [reply]
other problems:
signals are not complete in a tabular form, graphana has everything.
earlier all substations were with signals, then they disappeared.
now I will describe the problem more consistently, with screenshots.
p.s.
it is possible that errors occurred when adding missing parameters after importing to the Mongo database

**Comment by joyshmitz** [comment]
As well, we have in MONGO base some points:
https://drive.google.com/file/d/1nHpRbOvPLEyKpDBp_KlOE0YfJ2gb6SwX/view?usp=sharing

They are work properly and all was ok, but. now  those sygnals were gone from tabular viewer, from screen viewer:
https://drive.google.com/file/d/1m8k5Nsat7obWQt5mieuh7CyLfFGPcJ9O/view?usp=sharing
But we still can see subslation name in drop list, but not bay list, 
Also those points produce alarms in regular way:
https://drive.google.com/file/d/1Dp2ojz6_JQcSpQ7jY36x7aZ6zYfRMcQ6/view?usp=sharing

aslo still present in mongo realtime data and in grafana. What we can do to fix  screen and tabular viewer?

Also all of those points som time spontaniously set a zero values, so we have this breaking grafs:
https://drive.google.com/file/d/1g6SuXkPp94c3kSs4fc5f00ciWiKlitRg/view?usp=sharing
 
answer please

Also we have another issue...

**Reply by joyshmitz** [reply]
maybe its problem calculations.log
https://drive.google.com/file/d/18MVOJuA5hKUDKag5oeEYFRTxXctNEDd1/view?usp=sharing

2021/03/11 19:10:08.785668 is fault point, as we can see by graph:
https://drive.google.com/file/d/1xeLzhOk19pFcswTeP5LpGmNiKKReEku_/view?usp=sharing

**Reply by riclolsen** [reply]
Check properties "group1" and "group2" from tags that do not show.

Enter mongo shell (or use compass) and 

    db.realtimeData.find({_id: POINTKEYHERE }).pretty()

Copy and paste here for review.

Zero values must be coming from field probably, check protocol logs or errors on addressing (one point wrongly updating others). Check connection number, common address, object address.

"Read 0 points from MongoDB" This means there are no calculations defined.

The last problem is probably the same as the zeroed values. Try configuring one point at a time, begin with those problematic (remove all the other tags). For this, you can just make a mongo update to an unused connection number on tags and alter one by one to the original, like

    db.realtimeData.updateMany({protocolSourceConnectionNumber:61}, {$set:{protocolSourceConnectionNumber:-1}})

Then use Compass to alter one by one to the original.

**Reply by joyshmitz** [reply]
first query with point substitution, returned an empty string

https://drive.google.com/file/d/1qwSskhmmJUC8M_zmcEJC4MvQjkGXrIVM/view?usp=sharing

second problems 
  db.realtimeData.updateMany({protocolSourceConnectionNumber:1002}, {$set:{protocolSourceConnectionNumber:-1}})
{ acknowledged: true,
  insertedId: null,
  matchedCount: 0,
  modifiedCount: 0,
  upsertedCount: 0 }

**Reply by riclolsen** [reply]
Do first before queries

    use json_scada 

Or you be in the wrong database and all queries will return empty.

**Reply by joyshmitz** [reply]
thanks for the quick answers,
we found a parasitic process, `JSON_SCADA_demo_simul` was launched, causing confusion in the calculations

**Reply by joyshmitz** [reply]
Hi. We are solve the problem with demo stream for 104 protocol by deleting of demo simulator service. But now we have the same problem with  our own 104 streams. They are have same driver instance number, but different protocol connection numbers. They are working well, but sometimes point 41 from connection 1002 get the value from point 38 on connection #1. Point 40 get value from point 37 and others with this offset. What we can try to solve it?

**Reply by joyshmitz** [reply]
Those points have same protocolSourceAdress

Maybe where can be a problem?

**Reply by joyshmitz** [reply]
For some reason analog values does not produce alarm by hilimit and lolimit. How to set up them?

**Reply by riclolsen** [reply]
Do you mean protocolSourceObjectAddress?
No problem having multiple connections on the same instance of 104 client driver.
The code updates points using connection number, common address and object address.

        var filt =
            new rtFilt
            {
                protocolSourceConnectionNumber = iv.conn_number,
                protocolSourceCommonAddress = iv.common_address,
                protocolSourceObjectAddress = iv.address
            };
        ...
        listWrites.Add(new UpdateOneModel<rtData>(filt.ToBsonDocument(), update));
        
So, points on distinct connections should never interfere with each other, and with more reason points with distinct object addresses also should not interfere! You may look for duplicated object addresses on the same connection, use a query ordered by address like

    db.realtimeData.find({protocolSourceConnectionNumber: 1002}).projection({protocolSourceObjectAddress:1,tag:1}).sort({protocolSourceObjectAddress:1})

Can you send your connections list and the points that are interfering? (use Compass to copy documents).
You can also export your collections using Compass and send me via email.

**Reply by riclolsen** [reply]
> For some reason analog values does not produce alarm by hilimit and lolimit. How to set up them?

It should produce alarm when value is updated out of bounds.
Please include the points with problem in the list for review.

**Reply by joyshmitz** [reply]
send to e-mail realtimeData.json

**Comment by joyshmitz** [comment]
For some reasons we have wrong grafs for grafana  and internal graph service. as example, we have zero value of voltage, and it goes less then one minute, then we have some values, 10V as ex. We can see graph line in red, how it must look like. 

https://drive.google.com/file/d/1d0FbmzNAkP1okb7SkSOxHdjco_yaZSAT/view?usp=sharing

But we can see only lines from min and max changes

**Reply by riclolsen** [reply]
Edit panel, go to Display settings, set Staircase option. Or use bars or points instead of lines.

**Comment by riclolsen** [comment]
I have changed 104 client driver to include an "originator" property on "sourceDataUpdate". This way you can check what driver/connection is updating points. 

Stop all processes and, please replace your entire c:\json-scada\bin folder with

https://drive.google.com/file/d/1VYyl_oSe90DdRYZ0J7ZhMr-8QipxIVgy/view?usp=sharing

Also, install the latest dotnet core runtime.

https://dotnet.microsoft.com/download/dotnet/thank-you/runtime-5.0.4-windows-x64-installer

Activate processes again, wait for a wrong point update, then stop 104 client immediately. Chech the point with problem in Compass, see sourceDataUpdate object, originator property, should have a value like "IEC60870-5-104|1002" (driver_name|connection_number), check if driver and connection are correct.

**Reply by joyshmitz** [reply]
Did everything according to the instructions, the problem did not go away
![2021-03-18.png](https://user-images.githubusercontent.com/11138019/111672990-2de2c000-8823-11eb-9b2b-5d82947ddeb1.png)
value must be 0
it is received from stream # 1
these 9.954 ...
but it shows that I got from the right place
1002
but this is not true....


raised the scada on another computer, everything works without errors, copied to the server, errors.
Let's try to reinstall Windows, and Scada.
the server was shutting down abnormally at night, the electricity was turned off. maybe something went wrong

**Reply by riclolsen** [reply]
This means the driver and connection are right, value must be coming from protocol.
You can fire "iec104client 1 3" from command line (first stop the service) to see full protocol logs, disable connections not important. Watch for values with object address 40 and common address 1 in this connection.

**Reply by joyshmitz** [reply]
solved the problem by reinstalling the scada

**Comment by joyshmitz** [comment]
remove the limit of 500 entries in the event viewer 
this is to increase the parameter in the file events.html  - "g_nummaxevetr: 500, // número máximo de eventos em tempo real"

and format of presentation of date and time

**Reply by riclolsen** [reply]
For now, change here c:\json-scada\src\htdocs\events.html, line 75

  var WebSAGE =
  {
  g_nummaxevehist: 2500, // número máximo de eventos buscados do histórico
  g_nummaxevetr: 500, // número máximo de eventos em tempo real

 I will move these configuration options to config_viewers.js.

**Reply by joyshmitz** [reply]
super, thank you

**Reply by joyshmitz** [reply]
how to set date format to 24 hours and ddmmyyyy
date display format is set to scada or is it data base formatting

**Reply by riclolsen** [reply]
The functions used in Events/Tabular/Alarms viewers are 

    https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleString
    https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleTimeString
    https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleDateString

They are called with no arguments, so they convert to default browser locale definitions. Database formats should not interfere.

This can be changed on each client browser

    https://help.gofmx.com/hc/en-us/articles/115001796651-Changing-Date-Format-to-dd-mm-yy

I will create config options to force locale passing arguments to these functions (leaving defaults to the client browser as it is now).

**Reply by joyshmitz** [reply]
thanks for answer

**Comment by joyshmitz** [comment]
error in screen view, how to fix
![2021-03-21.png](https://user-images.githubusercontent.com/11138019/111899147-59d49000-8a33-11eb-906e-3a3810ea4c4a.png)

**Reply by joyshmitz** [reply]
We need function of general interrogation of 104 protocol instances, when we do accept action at alarm tab.
Is it in scada, or how to do it?

**Reply by riclolsen** [reply]
"error in screen view, how to fix"
This error is specific to a display file, I need the SVG file to locate the problem.
"We need function of general interrogation of 104 protocol instances, when we do accept action at alarm tab.
Is it in scada, or how to do it?"
Do you mean from the Alarms Viewer? The Ack button?
This not possible now. I can change the driver to accept commands for firing interrogations. Then we will need some kind of script linked to the desired action that will queue the interrogation commands.

**Reply by riclolsen** [reply]
Revisiting the code, I think the interrogation commands are already possible now, I will make sure to test.
I think the firing can be accomplished with a simple nodejs script tapping changes on userActions, when detected a change it will queue the necessary commands. I will write a template script for that and include it in the project. This will be a powerful feature.

**Reply by joyshmitz** [reply]
"error in screen view, how to fix" - won by cleaning the svg file from garbage

**Comment by joyshmitz** [comment]
The filter does not work in the alarms tab,![2021-03-21-1.png](https://user-images.githubusercontent.com/11138019/111929534-42db7f80-8abf-11eb-9fce-48d95d58dce9.png)
in the upper right corner draws an error in the activity field, the characters / * - E are replaced![2021-03-21-2.png](https://user-images.githubusercontent.com/11138019/111929650-a36abc80-8abf-11eb-8d94-b33183f952df.png)

**Reply by joyshmitz** [reply]
We need to write points value on realtime database only if it was changed, if the same with previous, do not write, not produce alarms again, (some sporadic mode)  and when we ack the point, need to clear compare list with previous values. Because we have big database size with same values, so we can't read the station values as often as we need, because event count for alarmed points gets fantastic values.

**Reply by riclolsen** [reply]
Look for F12 console errors on the alarms viewer, report if found any.

Protocol drivers will write every data update received, it does not control previous values, but these updates are always written to sourceDataUpdate property for later processing on cs_data_processor.js module, this module will only update the main value property and process alarms when the value is changed (or the quality changes).

**Comment by joyshmitz** [comment]
How to set up an event point to not produce event in events viewer if point value is 0. Now it produce event in any state

**Reply by joyshmitz** [reply]
please tell me, how to expand the time in the event view in the format hour-minute-second-millisecond

**Reply by riclolsen** [reply]
alarmState=1
alarmState [Double] - Considered state for alarm (0=off=false, 1=on=true, 2=both states, 3=state OFF->ON transition, -1=no state produces alarms but alarms can be signaled by other means) when type=digital.

**Reply by riclolsen** [reply]
To change date/time format in events viewer it is necessary the latest updates.

Uncomment the section below in c:\json-scada\src\htdocs\conf\config-viewers.js.

  // locale and time zone for date/time formatting (leave empty for browser defaults)
  // see https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleString
  var EventsViewer_LocaleTime = "en-US"; // locale for time format e.g "en-US"
  var EventsViewer_LocaleDate = "en-US"; // locale for date format e.g "en-US"
  // for options to work, a locale must be set
  var EventsViewer_LocaleTimeOptions = {
      // timeZone: 'America/New_York'
  }; 
  var EventsViewer_LocaleDateOptions = {
      // weekday: 'long', year: 'numeric', month: 'long', day: 'numeric', timeZone: 'America/New_York'
  }; 
  
Choose a suitable locale, play with options if necessary. https://elijahmanor.com/blog/format-js-dates-and-times

**Reply by joyshmitz** [reply]
there may have been a misunderstanding about
`the event view in the format **hour-minute-second-millisecond**`

**Reply by joyshmitz** [reply]
the new feature the `discussion` does not meet all the needs for the dialogue around edits and software settings is more convenient to build into `issues`

the discussion is more like a chat.
Such a socialization, I apologize, I went down the wrong road from the very beginning discussing changes in software, got confused in the questions of the answers in the sequence ...

what do you think Ricardo?

**Reply by riclolsen** [reply]
I agree. I suggest the Issues section for bugs reporting and new feature requests each in a separate topic.
Discussions can be used for more generic questions/answers, topic separation is good for locating distinct subjects.
By the way, I have put out the binary release 0.8 with the latest updates.

**Reply by joyshmitz** [reply]
thanks, and see you in issue's ))
P.S.
.. going to download, a fix is awaiting))
